python/gps/utility/visualization.py

```diff
@@ -0,0 +1,141 @@
+                        )
+                    sample_end_effector = sample.get(END_EFFECTOR_POINTS)
+                    dists_to_target = np.nanmin(np.sqrt(np.sum((sample_end_effector[:, :3] - target_position.reshape(1, -1))**2, axis = 1)), axis = 0)
+                    if dists_to_target <= THRESH[experiment]:
+                        successes[k][seed, i] = 1.0
+    
+    success_rate_ioc = np.mean(successes[0], axis=0)
+    success_rate_sup = np.mean(successes[1], axis=0)
+    success_rate_demo = np.mean(successes[2], axis=0)
+    success_rate_oracle = np.mean(successes[3], axis=0)
+    print "ioc mean: " + repr(success_rate_ioc.mean())
+    print "sup mean: " + repr(success_rate_sup.mean())
+    print "demo mean: " + repr(success_rate_demo.mean())
+    print "oracle mean: " + repr(success_rate_oracle.mean())
+    print "ioc: " + repr(success_rate_ioc)
+    print "sup: " + repr(success_rate_sup)
+    print "demo: " + repr(success_rate_demo)
+    print "oracle: " + repr(success_rate_oracle)
+
+    plt.close('all')
+    fig = plt.figure(figsize=(8, 5))
+
+    subplt = plt.subplot()
+    subplt.plot(ioc_conditions, success_rate_ioc, '-rx')
+    subplt.plot(ioc_conditions, success_rate_sup, '-gx')
+    subplt.plot(ioc_conditions, success_rate_demo, '-bx')
+    subplt.plot(ioc_conditions, success_rate_oracle, '-yx')
+    ax = plt.gca()
+    ax.xaxis.set_ticks_position('bottom')
+    ax.yaxis.set_ticks_position('left')
+    ax.spines['right'].set_visible(False)
+    ax.spines['top'].set_visible(False)
+    ax.legend(['S3G', 'cost regr', 'RL policy', 'oracle'], loc='lower left')
+    plt.ylabel('Success Rate', fontsize=22)
+    plt.xlabel('Log Mass', fontsize=22, labelpad=-4)
+    plt.title("Generalization for 2-link reacher", fontsize=25)
+    plt.savefig(gps._data_files_dir + 'distribution_of_sample_conditions_average_curve.png')
+    plt.close('all')
+
+def visualize_samples(gps, N, agent_config, experiment='reacher'):
+    """
+    Compare samples between IOC and demo policies and visualize them in a plot.
+    Args:
+        gps: GPS object.
+        N: number of samples taken from the policy for comparison
+        config: Configuration of the agent to sample.
+        experiment: whether the experiment is peg, reacher or pointmass.
+    """
+    pol_iter = gps._hyperparams['algorithm']['iterations'] - 1
+    algorithm_ioc = gps.data_logger.unpickle(gps._data_files_dir + 'algorithm_itr_%02d' % pol_iter + '.pkl')
+    algorithm_sup = gps.data_logger.unpickle(gps._hyperparams['common']['supervised_exp_dir'] + 'data_files/algorithm_itr_%02d' % pol_iter + '.pkl')
+    algorithm_demo = gps.data_logger.unpickle(gps._hyperparams['common']['demo_exp_dir'] + 'data_files/algorithm_itr_09.pkl') # Assuming not using 4 policies
+    algorithm_oracle = gps.data_logger.unpickle(gps._hyperparams['common']['demo_exp_dir'] + 'data_files_oracle/algorithm_itr_09.pkl')
+    M = agent_config['conditions']
+    pol_ioc = algorithm_ioc.policy_opt.policy
```

Comment:
why not make a list of algorithms, then use a list iteration to make a list of policies -- policies = [alg.policy_opt.policy for alg in algorithms]